File: macos_hid_diag/hid_test.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <IOKit/hid/IOHIDManager.h>
#include <IOKit/hid/IOHIDDevice.h>
#include <IOKit/IOKitLib.h>
#include <CoreFoundation/CoreFoundation.h>

static int g_callback_count = 0;
static uint8_t g_report_buf[64];

void input_report_callback(void *context, IOReturn result, void *sender,
                            IOHIDReportType type, uint32_t reportID,
                            uint8_t *report, CFIndex reportLength)
{
    g_callback_count++;
    printf("  [CALLBACK #%d] type=%d id=%d len=%ld data=", 
           g_callback_count, type, reportID, (long)reportLength);
    for (CFIndex i = 0; i < (reportLength < 8 ? reportLength : 8); i++)
        printf("%02x", g_report_buf[i]);
    printf("\n");
    fflush(stdout);
}

int main(void) {
    uint64_t entry_id = 4295026141ULL;
    
    CFDictionaryRef matching = IORegistryEntryIDMatching(entry_id);
    if (!matching) { fprintf(stderr, "IORegistryEntryIDMatching failed\n"); return 1; }
    
    io_service_t service = IOServiceGetMatchingService(kIOMainPortDefault, matching);
    printf("service_t: 0x%08x\n", service);
    if (!service) { fprintf(stderr, "Service not found\n"); return 1; }
    
    IOHIDDeviceRef dev = IOHIDDeviceCreate(kCFAllocatorDefault, service);
    IOObjectRelease(service);
    printf("IOHIDDeviceRef: %p\n", dev);
    if (!dev) { fprintf(stderr, "IOHIDDeviceCreate failed\n"); return 1; }
    
    IOReturn ret = IOHIDDeviceOpen(dev, kIOHIDOptionsTypeNone);
    printf("IOHIDDeviceOpen: 0x%08x (%s)\n", ret, ret == kIOReturnSuccess ? "OK" : "FAIL");
    if (ret != kIOReturnSuccess) { CFRelease(dev); return 1; }
    
    CFRunLoopRef rl = CFRunLoopGetCurrent();
    IOHIDDeviceScheduleWithRunLoop(dev, rl, kCFRunLoopDefaultMode);
    printf("Scheduled with run loop\n");
    
    IOHIDDeviceRegisterInputReportCallback(dev, g_report_buf, sizeof(g_report_buf),
                                           input_report_callback, NULL);
    printf("Callback registered\n");
    
    // Send OP_INIT frame
    uint8_t frame[64] = {0x10, 0x02, 0x00, 0x01, 0x01, 0x10, 0x10, 0x03, 0x11};
    ret = IOHIDDeviceSetReport(dev, kIOHIDReportTypeOutput, 0, frame, sizeof(frame));
    printf("IOHIDDeviceSetReport: 0x%08x\n", ret);
    
    // Pump run loop for 2 seconds
    printf("Pumping run loop for 2 seconds...\n");
    int iter = 0;
    CFAbsoluteTime deadline = CFAbsoluteTimeGetCurrent() + 2.0;
    while (CFAbsoluteTimeGetCurrent() < deadline) {
        SInt32 result = (SInt32)CFRunLoopRunInMode(kCFRunLoopDefaultMode, 0.05, true);
        iter++;
        if (result == kCFRunLoopRunHandledSource) {
            printf("  [iter %d] source handled!\n", iter);
        } else if (result == kCFRunLoopRunFinished) {
            printf("  [iter %d] run loop finished (no sources)!\n", iter);
        }
        if (g_callback_count > 0) break;
    }
    
    printf("Done. iter=%d callbacks=%d\n", iter, g_callback_count);
    
    IOHIDDeviceUnscheduleFromRunLoop(dev, rl, kCFRunLoopDefaultMode);
    IOHIDDeviceClose(dev, kIOHIDOptionsTypeNone);
    CFRelease(dev);
    return 0;
}

File: macos_hid_diag/hid_test.swift
import Foundation
import IOKit
import IOKit.hid

let VID = 0x0483
let PID = 0x5750
let REPORT_SZ = 64

func buildFrame(_ cmd: UInt8) -> [UInt8] {
    let n = UInt8(1)
    var chk = n ^ cmd
    var f: [UInt8] = [0x10, 0x02, 0x00, 0x01, n, cmd, 0x10, 0x03, chk]
    while f.count < REPORT_SZ { f.append(0) }
    return f
}

let mgr = IOHIDManagerCreate(kCFAllocatorDefault, IOOptionBits(kIOHIDOptionsTypeNone))
let matchDict = [kIOHIDVendorIDKey: VID, kIOHIDProductIDKey: PID] as CFDictionary
IOHIDManagerSetDeviceMatching(mgr, matchDict)

var openRet = IOHIDManagerOpen(mgr, IOOptionBits(kIOHIDOptionsTypeNone))
print("IOHIDManagerOpen: \(openRet == kIOReturnSuccess ? "OK" : "FAILED \(String(format:"0x%08x",UInt32(bitPattern:openRet)))")")
if openRet != kIOReturnSuccess { exit(1) }

// Give time for enumeration
IOHIDManagerScheduleWithRunLoop(mgr, CFRunLoopGetCurrent(), CFRunLoopMode.defaultMode.rawValue)
CFRunLoopRunInMode(.defaultMode, 0.5, false)

let devSetOpt = IOHIDManagerCopyDevices(mgr)
guard let devSet = devSetOpt else {
    print("No devices found")
    exit(1)
}
let count = CFSetGetCount(devSet)
print("Found \(count) device(s)")
guard count > 0 else { exit(1) }

var devPtr: UnsafeRawPointer? = nil
CFSetGetValues(devSet, &devPtr)
let dev = Unmanaged<IOHIDDevice>.fromOpaque(devPtr!).takeUnretainedValue()

if let prod = IOHIDDeviceGetProperty(dev, kIOHIDProductKey as CFString) {
    print("Product: \(prod)")
}

let devOpenRet = IOHIDDeviceOpen(dev, IOOptionBits(kIOHIDOptionsTypeNone))
print("IOHIDDeviceOpen: \(devOpenRet == kIOReturnSuccess ? "OK" : "FAILED \(String(format:"0x%08x",UInt32(bitPattern:devOpenRet)))")")
if devOpenRet != kIOReturnSuccess { exit(1) }

IOHIDDeviceScheduleWithRunLoop(dev, CFRunLoopGetCurrent(), CFRunLoopMode.defaultMode.rawValue)

var reportBuf = [UInt8](repeating: 0, count: REPORT_SZ)
IOHIDDeviceRegisterInputReportCallback(dev, &reportBuf, CFIndex(REPORT_SZ),
    { (ctx, result, sender, rtype, rid, report, rlen) in
        var hex = ""
        for i in 0..<min(Int(rlen), 16) { hex += String(format: "%02x ", report[i]) }
        print("  *** CALLBACK: len=\(rlen) \(hex)")
    }, nil)
print("Callback registered\n")

let cmds: [(String, UInt8)] = [("OP_INIT 0x10", 0x10), ("OP_FW 0x13", 0x13), ("OP_POLL 0x40", 0x40)]
for (label, cmd) in cmds {
    print("→ \(label)")
    let frame = buildFrame(cmd)
    let ret = IOHIDDeviceSetReport(dev, kIOHIDReportTypeOutput, CFIndex(0), frame, CFIndex(REPORT_SZ))
    if ret != kIOReturnSuccess {
        print("  SetReport FAILED: \(String(format:"0x%08x",UInt32(bitPattern:ret)))")
    } else {
        print("  SetReport OK, pumping 1.5s...")
        CFRunLoopRunInMode(.defaultMode, 1.5, false)
    }
    print()
}

IOHIDDeviceUnscheduleFromRunLoop(dev, CFRunLoopGetCurrent(), CFRunLoopMode.defaultMode.rawValue)
IOHIDDeviceClose(dev, IOOptionBits(kIOHIDOptionsTypeNone))
print("Done.")
